packages/compass-components/src/components/guide-cue/guide-cue.tsx

```diff
@@ -0,0 +1,200 @@
+import React, {
+  useEffect,
+  useCallback,
+  useRef,
+  useMemo,
+  useState,
+} from 'react';
+import { guideCueService, type ShowCueEventDetail } from './guide-cue-service';
+import { GuideCue as LGGuideCue } from '@leafygreen-ui/guide-cue';
+import type { GroupName } from './guide-cue-groups';
+
+const getDataCueId = ({
+  cueId,
+  groupId,
+}: {
+  cueId: string;
+  groupId?: string;
+}): string => {
+  return `guide-cue-${cueId}${groupId ? `-${groupId}` : ''}`;
+};
+
+type LGGuideCueProps = React.ComponentProps<typeof LGGuideCue>;
+
+type GroupAndStep =
+  | {
+      groupId: GroupName;
+      step: number;
+    }
+  | {
+      groupId?: never;
+      step?: never;
+    };
+
+// omit the props we are handling
+export type GuideCueProps<T> = Omit<
+  LGGuideCueProps,
+  | 'currentStep'
+  | 'refEl'
+  | 'numberOfSteps'
+  | 'onDismiss'
+  | 'open'
+  | 'onPrimaryButtonClick'
+  | 'setOpen'
+> &
+  GroupAndStep & {
+    cueId: string;
+    trigger: ({
+      ref,
+    }: {
+      ref: React.MutableRefObject<T | null>;
+    }) => React.ReactElement;
+  };
+
+export const GuideCue = <T extends HTMLElement>({
+  children,
+  trigger,
+  cueId,
+  groupId,
+  step,
+  ...restOfCueProps
+}: React.PropsWithChildren<GuideCueProps<T>>) => {
+  const [isOpen, setIsOpen] = useState(false);
+  const intersectionRef = useRef(true);
+  const refEl = useRef<T | null>(null);
+
+  const cueData = useMemo(() => {
+    if (!groupId) {
+      return { cueId, step: 1 };
+    }
+    return { cueId, groupId, step };
+  }, [cueId, groupId, step]);
+
+  const observerCallback = useCallback(
+    (entries: IntersectionObserverEntry[]) => {
+      const [entry] = entries;
+      intersectionRef.current = entry.isIntersecting;
+
+      guideCueService.onCueIntersectionChange(
+        entry.isIntersecting,
+        cueData.cueId,
+        cueData.groupId
+      );
+
+      if (!entry.isIntersecting && isOpen) {
+        setIsOpen(false);
+      }
```

Comment:
You can remove this isOpen check from there I think (unless I'm missing something), this creates a dependency in this callback that can be avoided, meaning that your observer in the effect below is also not recreated everytime isOpen changes

Reply:
nice catch. will fix it